File: src/c47/distributions/negBinom.c
/* This file is part of 43S.
 *
 * 43S is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * 43S is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with 43S.  If not, see <http://www.gnu.org/licenses/>.
 */

/********************************************//**
 * \file negBinom.c
 ***********************************************/

#include "distributions/negBinom.h"

#include "constantPointers.h"
#include "distributions/chi2.h"
#include "distributions/f.h"
#include "distributions/poisson.h"
#include "error.h"
#include "flags.h"
#include "mathematics/comparisonReals.h"
#include "mathematics/cpyx.h"
#include "mathematics/wp34s.h"
#include "registers.h"
#include "registerValueConversions.h"

#include "c47.h"


#if defined(SAVE_SPACE_DM42_15)
  void fnNegBinomialP  (uint16_t unusedButMandatoryParameter){}
  void fnNegBinomialL  (uint16_t unusedButMandatoryParameter){}
  void fnNegBinomialR  (uint16_t unusedButMandatoryParameter){}
  void fnNegBinomialI  (uint16_t unusedButMandatoryParameter){}
  void pdf_NegBinomial (const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext){}
  void cdfu_NegBinomial(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext){}
  void cdf_NegBinomial (const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext){}
  void cdf_NegBinomial2(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext){}
  void qf_NegBinomial  (const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext){}

#else
  static bool_t checkParamNegBinom(real_t *x, real_t *i, real_t *j) {
    if (!getRegisterAsReal(REGISTER_X, x)
        || !getRegisterAsReal(REGISTER_P, i)
        || !getRegisterAsReal(REGISTER_N, j))
      goto err;

    if(!checkRegisterNoFP(j)) {
      displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamNegBinom:", "r is not an integer", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
    else if(realIsNegative(x)) {
      displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamNegBinom:", "cannot calculate for x < 0", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
    else if(realIsZero(i) || realIsNegative(i) || realCompareGreaterEqual(i, const_1) || realIsZero(j) || realIsNegative(j)) {
      displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
      #if(EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamNegBinom:", "the parameters must be 0 < p < 1 and r > 0", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
    return true;

  err:
    if(getSystemFlag(FLAG_SPCRES)) {
      convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
    }
    return false;
  }


  void fnNegBinomialP(uint16_t unusedButMandatoryParameter) {
    real_t val, ans, prob, num;

    if(!saveLastX()) {
      return;
    }

    if(checkParamNegBinom(&val, &prob, &num)) {
      if(realIsAnInteger(&val)) {
        pdf_NegBinomial(&val, &prob, &num, &ans, &ctxtReal39);
      }
      else {
        realZero(&ans);
      }
      if(realIsNaN(&ans)) {
        displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fnNegBinomialP:", "a parameter is invalid", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      }
      else {
        reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
        convertRealToReal34ResultRegister(&ans, REGISTER_X);
      }
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }


  void fnNegBinomialL(uint16_t unusedButMandatoryParameter) {
    real_t val, ans, prob, num;

    if(!saveLastX()) {
      return;
    }

    if(checkParamNegBinom(&val, &prob, &num)) {
      cdf_NegBinomial(&val, &prob, &num, &ans, &ctxtReal39);
      if(realIsNaN(&ans)) {
        displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fnNegBinomialL:", "a parameter is invalid", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      }
      else {
        reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
        convertRealToReal34ResultRegister(&ans, REGISTER_X);
      }
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }


  void fnNegBinomialR(uint16_t unusedButMandatoryParameter) {
    real_t val, ans, prob, num;

    if(!saveLastX()) {
      return;
    }

    if(checkParamNegBinom(&val, &prob, &num)) {
      cdfu_NegBinomial(&val, &prob, &num, &ans, &ctxtReal39);
      if(realIsNaN(&ans)) {
        displayDomainErrorMessage(ERROR_INVALID_DISTRIBUTION_PARAM, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fnNegBinomialR:", "a parameter is invalid", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      }
      else {
        reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
        convertRealToReal34ResultRegister(&ans, REGISTER_X);
      }
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }


  void fnNegBinomialI(uint16_t unusedButMandatoryParameter) {
    real_t val, ans, prob, num;

    if(!saveLastX()) {
      return;
    }

    if(checkParamNegBinom(&val, &prob, &num)) {
      if(realCompareLessEqual(&val, const_0) || realCompareGreaterEqual(&val, const_1)) {
        displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fnNegBinomialI:", "the argument must be 0 < x < 1", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
        if(getSystemFlag(FLAG_SPCRES)) {
          convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
        }
        return;
      }
      qf_NegBinomial(&val, &prob, &num, &ans, &ctxtReal39);
      if(realIsNaN(&ans)) {
        displayDomainErrorMessage(ERROR_NO_ROOT_FOUND, ERR_REGISTER_LINE, REGISTER_X);
        #if(EXTRA_INFO_ON_CALC_ERROR == 1)
          moreInfoOnError("In function fnNegBinomialI:", "WP34S_Qf_Binomial did not converge", NULL, NULL);
        #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
        if(getSystemFlag(FLAG_SPCRES)) {
          convertRealToReal34ResultRegister(const_NaN, REGISTER_X);
        }
        return;
      }
      reallocateRegister(REGISTER_X, dtReal34, REAL34_SIZE_IN_BLOCKS, amNone);
      convertRealToReal34ResultRegister(&ans, REGISTER_X);
    }

    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }


  bool_t negBinom_param(const real_t *r, real_t *res) {
    if(realIsSpecial(r)) {
      realCopy(const_NaN, res);
      return false;
    }
    if((!realIsPositive(r)) || (!realIsAnInteger(r))) {
      realCopy(const_0, res);
      return false;
    }
    return true;
  }

  // PDF[NB(r, p)](k) = [(k + r - 1) C k] p^k (1 - p)^r
  void pdf_NegBinomial(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext) {
    real_t p, q, xx, c;

    if(!negBinom_param(r, res)) {
      return;
    }
    if(realIsNegative(x)) {
      realZero(res);
      return;
    }

    realMultiply(p0, const__1, &p, realContext);
    WP34S_Ln1P(&p, &p, realContext);       // ln(1 - p0)
    realMultiply(&p, r, &p, realContext);  // ln((1 - p0) ^ r)

    WP34S_Ln(p0, &q, realContext);
    realMultiply(&q, x, &q, realContext);  // ln(p0 ^ x)
    realAdd(&p, &q, &p, realContext);

    realAdd(x, r, &q, realContext);
    realSubtract(&q, const_1, &q, realContext);
    realCopy(x, &xx);
    logCyxReal(&q, &xx, &c, realContext);

    realAdd(&p, &c, &p, realContext);
    realExp(&p, res, realContext);
  }

  // I[p](k, r)
  void cdfu_NegBinomial(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext) {
    real_t p;

    if(!negBinom_param(r, res)) {
      return;
    }
    realToIntegralValue(x, &p, DEC_ROUND_CEILING, realContext);
    if(realCompareLessThan(&p, const_1)) {
      realCopy(const_1, res);
      return;
    }

    WP34S_betai(r, &p, p0, res, realContext);
  }

  // 1 - I[p](k + 1, r)
  void cdf_NegBinomial(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext) {
    real_t p;

    if(!negBinom_param(r, res)) {
      return;
    }
    realToIntegralValue(x, &p, DEC_ROUND_FLOOR, realContext);
    cdf_NegBinomial2(&p, p0, r, res, realContext);
  }

  void cdf_NegBinomial2(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext) {
    real_t p, q;

    if(realCompareLessThan(x, const_0)) {
      realZero(res);
      return;
    }

    realSubtract(const_1, p0, &p, realContext);

    realAdd(x, const_1, &q, realContext);
    WP34S_betai(&q, r, &p, res, realContext);
  }

  void qf_NegBinomial(const real_t *x, const real_t *p0, const real_t *r, real_t *res, realContext_t *realContext) {
    real_t p0c, pr, mean, var, s;

    if(!negBinom_param(r, res)) {
      return;
    }
    if(realCompareLessEqual(x, const_0)) {
      realZero(res);
      return;
    }

    realSubtract(const_1, p0, &p0c, realContext); // 1 - p
    realMultiply(p0, r, &pr, realContext);
    realDivide(&pr, &p0c, &mean, realContext);    // mean = pr/(1-p)

    realDivide(&mean, &p0c, &var, realContext);   // variance = pr/((1-p)^2)
    realSquareRoot(&var, &var, realContext);

    WP34S_normal_moment_approx(x, &var, &mean, &s, realContext);
    WP34S_Qf_Newton(QF_NEWTON_NEGBINOM, x, &s, p0, r, NULL, res, realContext);
  }

#endif //SAVE_SPACE_DM42_15

File: zig_src/shell/distributions/neg_binom.zig
// SPDX-License-Identifier: GPL-3.0-only
//
// Zig owner for src/c47/distributions/negBinom.c: the Negative Binomial
// distribution commands (fnNegBinomialP/L/R/I) and the WP34S math borrowings.
// Part of the SAVE_SPACE_DM42_17 cluster, so compiled only on host and DMCP5 and
// stubbed on every DM42 package via strip_17.
//
// cdf_NegBinomial2 and pdf_NegBinomial are exported with C linkage because
// still-C cluster members reach them where the cluster is kept (the geometric
// qf_discrete_final dispatcher and the f-distribution Newton solver's NEGBINOM
// case; the f distribution's PDF path). The quantile uses the shared C Newton
// solver (WP34S_Qf_Newton), the regularised incomplete beta (WP34S_betai),
// logCyxReal, and the poisson owner's WP34S_normal_moment_approx, all bound
// through frontier_distribution_runtime. The testSuite drives every branch via
// tests/negBinomial_{p,l,r,i}.txt.

const dr = @import("distribution_runtime.zig");
pub const real_t = dr.real_t;
pub const realContext_t = dr.realContext_t;

// WP34S_Qf_Newton distribution selector (defines.h).
const QF_NEWTON_NEGBINOM: u32 = 4;

fn checkParamNegBinom(x: *real_t, i: *real_t, j: *real_t) bool {
    if (!dr.saveLastX()) {
        return false;
    }

    if (!dr.getRegisterAsReal(dr.REGISTER_X, x) or !dr.getRegisterAsReal(dr.REGISTER_P, i) or !dr.getRegisterAsReal(dr.REGISTER_N, j)) {
        dr.specialResultNaN();
        return false;
    }

    if (!dr.checkRegisterNoFP(j)) {
        dr.displayDomainErrorMessage(dr.ERROR_INVALID_DISTRIBUTION_PARAM, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamNegBinom:", "r is not an integer", null, null);
        dr.specialResultNaN();
        return false;
    } else if (dr.realIsNegative(x)) {
        dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamNegBinom:", "cannot calculate for x < 0", null, null);
        dr.specialResultNaN();
        return false;
    } else if (dr.realIsZero(i) or dr.realIsNegative(i) or dr.realCompareGreaterEqual(i, dr.const1()) or dr.realIsZero(j) or dr.realIsNegative(j)) {
        dr.displayDomainErrorMessage(dr.ERROR_INVALID_DISTRIBUTION_PARAM, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamNegBinom:", "the parameters must be 0 < p < 1 and r > 0", null, null);
        dr.specialResultNaN();
        return false;
    }
    return true;
}

// Tail shared by P/L/R: store the answer, or report an invalid parameter on NaN.
fn storeOrInvalid(ans: *const real_t, comptime where: [*:0]const u8) void {
    if (dr.realIsNaN(ans)) {
        dr.displayDomainErrorMessage(dr.ERROR_INVALID_DISTRIBUTION_PARAM, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError(where, "a parameter is invalid", null, null);
    } else {
        dr.convertRealToResultRegister(ans, dr.REGISTER_X, dr.amNone);
    }
    dr.adjustResult(dr.REGISTER_X, false, false, dr.REGISTER_X, -1, -1);
}

pub fn negBinomialP(unused_but_mandatory_parameter: u16) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var ans: real_t = undefined;
    var prob: real_t = undefined;
    var num: real_t = undefined;

    if (checkParamNegBinom(&val, &prob, &num)) {
        if (dr.isAnInteger(&val)) {
            pdfNegBinomial(&val, &prob, &num, &ans, &dr.ctxtReal39);
        } else {
            dr.setZero(&ans);
        }
        storeOrInvalid(&ans, "In function fnNegBinomialP:");
    }
}

pub fn negBinomialL(unused_but_mandatory_parameter: u16) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var ans: real_t = undefined;
    var prob: real_t = undefined;
    var num: real_t = undefined;

    if (checkParamNegBinom(&val, &prob, &num)) {
        cdfNegBinomial(&val, &prob, &num, &ans, &dr.ctxtReal39);
        storeOrInvalid(&ans, "In function fnNegBinomialL:");
    }
}

pub fn negBinomialR(unused_but_mandatory_parameter: u16) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var ans: real_t = undefined;
    var prob: real_t = undefined;
    var num: real_t = undefined;

    if (checkParamNegBinom(&val, &prob, &num)) {
        cdfuNegBinomial(&val, &prob, &num, &ans, &dr.ctxtReal39);
        storeOrInvalid(&ans, "In function fnNegBinomialR:");
    }
}

pub fn negBinomialI(unused_but_mandatory_parameter: u16) void {
    _ = unused_but_mandatory_parameter;
    var val: real_t = undefined;
    var ans: real_t = undefined;
    var prob: real_t = undefined;
    var num: real_t = undefined;

    if (checkParamNegBinom(&val, &prob, &num)) {
        if (dr.realCompareLessEqual(&val, dr.const0()) or dr.realCompareGreaterEqual(&val, dr.const1())) {
            dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
            dr.moreInfoOnError("In function fnNegBinomialI:", "the argument must be 0 < x < 1", null, null);
            dr.specialResultNaN();
            return;
        }
        qfNegBinomial(&val, &prob, &num, &ans, &dr.ctxtReal39);
        if (dr.realIsNaN(&ans)) {
            dr.displayDomainErrorMessage(dr.ERROR_NO_ROOT_FOUND, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
            dr.moreInfoOnError("In function fnNegBinomialI:", "WP34S_Qf_Binomial did not converge", null, null);
            dr.specialResultNaN();
            return;
        }
        dr.convertRealToResultRegister(&ans, dr.REGISTER_X, dr.amNone);
        dr.adjustResult(dr.REGISTER_X, false, false, dr.REGISTER_X, -1, -1);
    }
}

// ---------------------------------------------------------------------------
// WP34S math borrowings (upstream comment: "borrowed from the WP34S project").
// ---------------------------------------------------------------------------

fn negBinomParam(r: *const real_t, res: *real_t) bool {
    if (dr.realIsSpecial(r)) {
        dr.setNaN(res);
        return false;
    }
    if (!dr.realIsPositive(r) or !dr.isAnInteger(r)) {
        dr.setZero(res);
        return false;
    }
    return true;
}

// PDF[NB(r, p)](k) = [(k + r - 1) C k] p^k (1 - p)^r
pub fn pdfNegBinomial(x: *const real_t, p0: *const real_t, r: *const real_t, res: *real_t, ctx: *realContext_t) void {
    var p: real_t = undefined;
    var q: real_t = undefined;
    var xx: real_t = undefined;
    var c: real_t = undefined;

    if (!negBinomParam(r, res)) {
        return;
    }
    if (dr.realIsNegative(x)) {
        dr.setZero(res);
        return;
    }

    dr.realMinus(p0, &p, ctx);
    dr.WP34S_Ln1P(&p, &p, ctx); // ln(1 - p0)
    dr.realMultiply(&p, r, &p, ctx); // ln((1 - p0) ^ r)

    dr.WP34S_Ln(p0, &q, ctx);
    dr.realMultiply(&q, x, &q, ctx); // ln(p0 ^ x)
    dr.realAdd(&p, &q, &p, ctx);

    dr.realAdd(x, r, &q, ctx);
    dr.realSubtract(&q, dr.const1(), &q, ctx);
    dr.realCopy(x, &xx);
    dr.logCyxReal(&q, &xx, &c, ctx);

    dr.realAdd(&p, &c, &p, ctx);
    dr.realExponential(&p, res, ctx);
}

// I[p](k, r)
fn cdfuNegBinomial(x: *const real_t, p0: *const real_t, r: *const real_t, res: *real_t, ctx: *realContext_t) void {
    var p: real_t = undefined;

    if (!negBinomParam(r, res)) {
        return;
    }
    dr.toIntegralValue(x, &p, dr.DEC_ROUND_CEILING, ctx);
    if (dr.realLessThan(&p, dr.const1())) {
        dr.setOne(res);
        return;
    }
    dr.WP34S_betai(r, &p, p0, res, ctx);
}

// 1 - I[p](k + 1, r)
fn cdfNegBinomial(x: *const real_t, p0: *const real_t, r: *const real_t, res: *real_t, ctx: *realContext_t) void {
    var p: real_t = undefined;

    if (!negBinomParam(r, res)) {
        return;
    }
    dr.toIntegralValue(x, &p, dr.DEC_ROUND_FLOOR, ctx);
    cdfNegBinomial2(&p, p0, r, res, ctx);
}

pub fn cdfNegBinomial2(x: *const real_t, p0: *const real_t, r: *const real_t, res: *real_t, ctx: *realContext_t) void {
    var p: real_t = undefined;
    var q: real_t = undefined;

    if (dr.realLessThan(x, dr.const0())) {
        dr.setZero(res);
        return;
    }

    dr.realSubtract(dr.const1(), p0, &p, ctx);
    dr.realAdd(x, dr.const1(), &q, ctx);
    dr.WP34S_betai(&q, r, &p, res, ctx);
}

fn qfNegBinomial(x: *const real_t, p0: *const real_t, r: *const real_t, res: *real_t, ctx: *realContext_t) void {
    var p0c: real_t = undefined;
    var pr: real_t = undefined;
    var mean: real_t = undefined;
    var variance: real_t = undefined;
    var s: real_t = undefined;

    if (!negBinomParam(r, res)) {
        return;
    }
    if (dr.realCompareLessEqual(x, dr.const0())) {
        dr.setZero(res);
        return;
    }

    dr.realSubtract(dr.const1(), p0, &p0c, ctx); // 1 - p
    dr.realMultiply(p0, r, &pr, ctx);
    dr.realDivide(&pr, &p0c, &mean, ctx); // mean = pr/(1-p)

    dr.realDivide(&mean, &p0c, &variance, ctx); // variance = pr/((1-p)^2)
    dr.realSquareRoot(&variance, &variance, ctx);

    dr.WP34S_normal_moment_approx(x, &variance, &mean, &s, ctx);
    dr.WP34S_Qf_Newton(QF_NEWTON_NEGBINOM, x, &s, p0, r, null, res, ctx);
}
